lib/poison/decoder.ex

```diff
@@ -40,9 +40,13 @@ defmodule Poison.Decode do
   end
 
   defp transform_struct(value, _keys, as, options) do
-    kv = for k <- Map.keys(Map.from_struct(as.__struct__)) do
-      {k, Map.get(value, Atom.to_string(k))}
-    end
+    kv = Map.keys(Map.from_struct(as.__struct__))
+          |> Enum.reduce([], fn key, acc ->
+               case Map.get(value, Atom.to_string(key)) do
+                 nil -> acc
+                 value -> [{key, value} | acc]
+               end
+             end)
```

Comment:
Good point! In that case I would write it like this:

```elixir
Enum.into(as.__struct__, %{}, fn
  {:__struct__, struct_value} ->
    {:__struct__, struct_value}
  {key, struct_value} ->
    case Map.fetch(value, Atom.to_string(key)) do
      {:ok, json_value} -> {key, json_value}
      :error -> {key, struct_value}
    end
end)
|> Poison.Decoder.decode(options)
```

This code avoids calling `struct/2` so the input is not traversed twice. It also allows actual `null` values in the json to override defaults in the struct.